rts/motoko-rts/src/idl.rs

```diff
@@ -486,3 +512,335 @@ unsafe extern "C" fn skip_fields(tb: *mut Buf, buf: *mut Buf, typtbl: *mut *mut
+                    }
+                    advance = true;
+                }
+                return true;
+            }
+            (IDL_CON_variant, IDL_CON_variant) => {
+                let n1 = leb128_decode(&mut tb1);
+                let mut n2 = leb128_decode(&mut tb2);
+                for _ in 0..n1 {
+                    if n2 == 0 {
+                        break 'return_false;
+                    };
+                    let tag1 = leb128_decode(&mut tb1);
+                    let t11 = sleb128_decode(&mut tb1);
+                    let mut tag2: u32;
+                    let mut t21: i32;
+                    loop {
+                        tag2 = leb128_decode(&mut tb2);
+                        t21 = sleb128_decode(&mut tb2);
+                        n2 -= 1;
+                        if !(tag2 < tag1 && n2 > 0) {
+                            break;
+                        }
+                    }
+                    if tag1 != tag2 {
+                        break 'return_false;
+                    };
+                    if !sub(rel, p, typtbl1, typtbl2, end1, end2, t11, t21) {
+                        break 'return_false;
+                    }
+                }
+                return true;
+            }
+            (IDL_CON_service, IDL_CON_service) => {
+                let mut n1 = leb128_decode(&mut tb1);
+                let n2 = leb128_decode(&mut tb2);
+                for _ in 0..n2 {
+                    if n1 == 0 {
+                        break 'return_false;
+                    };
+                    let len2 = leb128_decode(&mut tb2);
+                    let p2 = tb2.ptr;
+                    Buf::advance(&mut tb2, len2);
+                    let t21 = sleb128_decode(&mut tb2);
+                    let mut len1: u32;
+                    let mut p1: *mut u8;
+                    let mut t11: i32;
+                    let mut cmp: i32;
+                    loop {
+                        len1 = leb128_decode(&mut tb1);
+                        p1 = tb1.ptr;
+                        Buf::advance(&mut tb1, len1);
+                        t11 = sleb128_decode(&mut tb1);
+                        n1 -= 1;
+                        cmp = utf8_cmp(len1, p1, len2, p2);
+                        if cmp < 0 && n1 > 0 {
+                            continue;
+                        };
+                        break;
+                    }
+                    if !(cmp == 0) {
+                        break 'return_false;
+                    };
+                    if !sub(rel, p, typtbl1, typtbl2, end1, end2, t11, t21) {
+                        break 'return_false;
+                    }
+                }
+                return true;
+            }
+            // default
+            (_, _) => {
+                break 'return_false;
+            }
+        }
+    }
+    // remember negative result ...
+    if t1 >= 0 && t2 >= 0 {
+        rel.disprove(p, t1 as u32, t2 as u32);
+    }
```

Comment:
Remembering negative results is “just” for performance; it would terminate and be correct without this, right?

Reply:
Yes but a vec might involved many repeated negative tests without failing so I think it's worth optimizing.

Reply:
Yes, I agree. Worth leaving comments many to highlight the different role of caching positive and negative results here, for our future selfs.

Reply:
You coming back?

Reply:
I worringly feel like I have never left

Reply:
That'll change...